R/final_size.R

```diff
@@ -1,54 +1,77 @@
 #' @examples
-#' final_size()
+#' library(socialmixr)
+#' data(polymod)
+#' contact_data <- contact_matrix(polymod, countries = "United Kingdom", age.limits = c(0,20,40))
+#' 
+#' c_matrix <-t(contact_data$matrix) # Define contact matrix (entry ij is contacts in group i reported by group j)
+#' d_vector <- contact_data$participants$proportion # Define proportion in each age group
+#' p_suscep <- c(1,0.5,0.5) # Define proportion of age group that is susceptible to infection
+#' 
+#' # Run final size model
+#' final_size(r0=2,contact_matrix = c_matrix, demography_vector = d_vector, prop_suscep = p_suscep)
+#'
+final_size <- function(r0 = 2, contact_matrix, demography_vector,
+                       prop_suscep = NULL) {
 
-final_size<-function(r0=2,contact_matrix,demography_vector,prop_suscep=NULL){
-  
   # Check inputs
-  if(is.null(prop_suscep)){prop_suscep <- demography_vector*0 + 1} # Assume fully susceptible if no entry
-  if(length(contact_matrix[1,])!=length(demography_vector)){stop("demography vector needs to be same size as contact matrix")}
-  if(length(demography_vector)!=length(prop_suscep)){stop("demography vector needs to be same size as susceptibility vector")}
-  pp0=as.numeric(demography_vector/sum(demography_vector))
-  
+  if (is.null(prop_suscep)) {
+    prop_suscep <- demography_vector * 0 + 1
+  } # Assume fully susceptible if no entry
+  if (length(contact_matrix[1, ]) != length(demography_vector)) {
+    stop("demography vector needs to be same size as contact matrix")
+  }
+  if (length(demography_vector) != length(prop_suscep)) {
+    stop("demography vector needs to be same size as susceptibility vector")
+  }
+  pp0 <- as.numeric(demography_vector / sum(demography_vector))
+
   # Scale next generation matrix so max eigenvalue=r0
   mm0 <- as.matrix(contact_matrix)
-  mm0 <- r0*mm0/max(Re(eigen(mm0)$values))
-  
+  mm0 <- r0 * mm0 / max(Re(eigen(mm0)$values))
+
   # Define transmission matrix A = mm_ij*pp_j/pp_i
-  beta1 <- (mm0* prop_suscep) /pp0
-  beta2 <- t(t(beta1)*pp0)
-  
+  beta1 <- (mm0 * prop_suscep) / pp0
+  beta2 <- t(t(beta1) * pp0)
+
   # Newton method for solving final size equation: A(1-x) = -log(x)
-  
+
   # Define functions f and f'
   vsize <- length(pp0)
-  f1 <- function(beta2,x){ beta2%*%(1-x) +log(x) }
-  f2 <- function(beta2,x,size){-beta2+diag(size)/x }
+  f1 <- function(beta2, x) {
+    beta2 %*% (1 - x) + log(x)
+  }
+  f2 <- function(beta2, x, size) {
+    -beta2 + diag(size) / x
+  }
 
   # Set storage vector and precision
   iterations <- 30
-  iterate_output <- matrix(NA,nrow=iterations,ncol=vsize)
-  x0 <- 0.001*pp0 # Set starting point
-  
-  for(ii in 1:iterations){
-    if(ii==1){
-        xx=x0
-        iterate_output[ii,]=xx
-      }else{
-        dx <- solve(f2(beta2,xx,vsize), -f1(beta2,xx))
-        iterate_output[ii,] <- xx+dx
-        xx <- as.numeric(xx+dx)
-      }
+  iterate_output <- matrix(NA, nrow = iterations, ncol = vsize)
+  x0 <- 0.001 * pp0 # Set starting point
```

Comment:
We may want to move hard-coded values to arguments of the function, defaulting to these values?

Reply:
Added as issue #22